### DISCUSSION TITLE
How do I render a template from a (non-block) helper?

### DISCUSSION THREAD

**Discussion by heemskerkerik** [Handlebars-Net/Handlebars.Net#444] [Q&A]
I'm trying to implement a front-end for Ghost, and one of the helpers they expose to their themes is `{{navigation}}`, which pulls a JSON object from settings, and feeds that to another template.

The basics are simple:

```csharp
var vm = new
         {
             navigation = new[]
                          {
                              new
                              {
                                  label = "Home",
                                  url = "/",
                              }
                          }
         };

var view = handlebars.CompileView("navigation.hbs");
using var writer = output.CreateWrapper();
view(
    writer,
    new
    {
        navigation = navigation,
    },
    // ???
);
```

But what do I pass as a third argument to the template to preserve the values in `options.Data`? I've tried passing `options.Frame`, but that results in the runtime crapping itself with

```
BadImageFormatException: An attempt was made to load a program with an incorrect format. (0x8007000B)
```

**Comment by oformaniuk** [comment]
I have not check my theory but it looks like you're using wrong `CompileView` overload. You should use a different one:
```csharp
HandlebarsTemplate<TextWriter, object, object> CompileView(string templatePath, ViewReaderFactory readerFactoryFactory)
```

For a reference you can look at the following test: [`CanLoadAWriterViewWithALayout`](https://github.com/Handlebars-Net/Handlebars.Net/blob/01ad0e69e56331371f85a1c59c008d7ecfd06f78/source/Handlebars.Test/ViewEngine/ViewEngineTests.cs#L34)

Regarding passing a `@data` object to a child template I'm not sure that it's possible/easy to do assuming current implementation as [`DataValues`](https://github.com/Handlebars-Net/Handlebars.Net/blob/master/source/Handlebars/ValueProviders/DataValues.cs) does not expose any information regarding existing values.
Without changing implementation you can access [`ContextDataObject`](https://github.com/Handlebars-Net/Handlebars.Net/blob/master/source/Handlebars/BindingContext.cs#L34) property of `options.Frame` using reflation and pass it's value as a third argument. However I'd probably not recommend using this approach for a long run as it may be broken in future updates.

A simple change to fix the problem can be made: expose `Keys` (or any better name) from [`DataValues`](https://github.com/Handlebars-Net/Handlebars.Net/blob/master/source/Handlebars/ValueProviders/DataValues.cs) to allow constructing new `@data` object when needed.

**Reply by oformaniuk** [reply]
@heemskerkerik , alternatively you can look at making `HandlebarsCompiler` public and use its `CompileView`. This can give you a better control and you'd be able to execute `navigation` view in child context omitting the need to pass `@data` object at all.
